### ISSUE TITLE
System.ServiceModel.Primitives does not seem to be compatible with .net 6

### ISSUE THREAD

**Issue by JonSaffron** [dotnet/wcf#5694]
Although NuGet says that the package is compatible with .net 6, and there is no problem bringing in the package, I cannot get Visual Studio to see anything in the System.ServiceModel namespace with the exception of the Syndication namespace.
![Image](https://github.com/user-attachments/assets/0440802c-c495-4071-ae9a-846844aecccc)

**To Reproduce**
Create a test project targeting framework net6.0.
Add package System.ServiceModel.Primitives 8.1.0
Note that none of the expected objects in the System.ServiceModel are available
Change the targeted framework to net8.0
Note that the expected objects now become available

**Expected behavior**
Either for nuget.org to clearly show that the package is NOT compatible with .net 6 (as it was with version 8.0.0),
OR for the package to be made compatible with .net 6

**Comment by mconnew** [comment]
The major version number of the WCF Client packages align with the minimum .NET version that is supports. So the 6.x packages were for .NET 6.0 and later, and the 8.x packages are for .NET 8.0 and later (ie .NET 9.0). The end of support date for the 6.x packages aligns with the end of support date of .NET 6.  

If you are targeting the now out of support .NET 6, you could use the now out of support WCF Client 6.x packages.  

The confusion you're seeing with what versions nuget says the package supports is due to adding back support for targeting netstandard2.0 in a library and being able to use that library on .NET Framework and .NET 8.0. To support this, we need to provide a reference assembly for netstandard2.0. This then gets picked up by any TFM which supports netstandard2.0, including .NET Framework, .NET 8, and any .NET/.NET Core version since .NET Core 2.0. If you target .NET Framework, you don't want to see the api's from the package, you want to see the api's from System.ServiceModel.dll, so we put an empty reference assembly in ref/net462. For .NET 8, we have a larger api surface (otherwise we could just target netstandard2.0), so that has its own reference assembly which includes some api's not available on .NET Framework. We then need lib files for each of these platforms, with the one for netstandard2.0 being a dummy empty implementation. To prevent picking up the netstandrd2.0 reference assembly on older unsupported target frameworks (such as net6.0), we placed a dummy folder at ref/netcoreapp2.0 with an empty placeholder file and no reference assemblies. If you're targeting anything from netcoreapp2.0 through to net7.0, this is the reference folder that will get picked up. This is to avoid someone adding the package to a project targeting an unsupported runtime such as net6.0, compiling against the netstandard2.0 reference assembly, potentially doing a lot of development work, and then failing at runtime due to not having an implementation.  

We can't stop nuget from reporting that the package supports a long list of tfm's which isn't actually supported as long as we have a reference assembly for netstandard2.0. And we have to have a reference assembly for netstandard2.0 to support targeting a library at netstandard2.0 and using it on .NET Framework and .NET 8 (as surprisingly large number of people do that and have had to keep using the 4.10.x release until we released the 8.1 release).  

The WCF packages aren't the only ones with this weird situation. If you take a look at System.Security.Cryptography.Xml, at the top of the page it lists .NET 8.0, .NET Standard 2.0, and .NET Framework 4.6.2. If you try to use the 9.0.0 package, they have a targets file in buildTransitive/netcoreapp2.0 which gets applied for anything that supports netcoreapp2.0 and above. They then have a dummy buildTransitive/net8.0 folder with nothing in it to suppress this target file on .NET 8 and .NET 9. What this target file does is create a build warning about using this down level. They can do this as they have an implementation that they are able to compile for netstandard2.0 and that's the version you would pick up when using it with net6.0. While theirs isn't a hard break, it is a big warning saying this is untested and potentially will break/do the wrong thing at runtime as it's untested. We can't do that, so instead opted to create a build break rather than have your compilation succeed and have a runtime breakage instead.  

We do support every supported .NET and .NET Framework 4.x version with our latest package. If there was a way to improve the messaging for nuget (and there isn't that I know of), it would be low priority as it would be work to prevent an out of support build environment from upgrading to newer packages. Basically supporting an out of support platform, which is a bit of an oxymoron.  

If you know of a way to improve how nuget behaves while maintaining the constraints of working for .NET Framework 4.6.2+, and all in support versions of .NET, without trying to make the implementation work on unsupported platforms (it prevents us adopting new features), feel free to open a PR.  

Unless something changes, there's nothing we can do here so closing the issue.

**Comment by aueelis** [comment]
@mconnew Thank you for this detailed explanation!

We were also targeting our packages for netstandard2.0, to support net472 and net8.0. Now, I changed all our packages, which (to the best of my knowledge) had full netstandard2.0 support to now only support those two target frameworks.

So what you are saying is, that you pretend to support netstandard2.0, so that others can pretend to support netstandard2.0, instead of forcing the correct migration to multi-targeting?

This would be quite disappointing, but at least I now know what to expect.

**Comment by mconnew** [comment]
It's a bit more nuanced than that. Every platform that supports netstandard2.0 except .NET [Core] ships with its own implementation of WCF in-box. For example, Xamarin was based on mono, and they shipped the mono implementation of System.ServiceModel.dll with a few changes to accommodate lack of Ref.Emit (they fixed that with their own client generator replacing svcutil/Add Service Reference which generated code which didn't need a real client proxy). If you look in the lib folder of the System.ServiceModel.Primitives package, you will see a MonoAndroid10 folder which is empty. This is so that when you include the nuget package in a library project and that gets included into a Xamarin project, it prevents our implementation from being deployed. The Xamarin toolchain has some mechanism to detect you are using WCF packages and provides their own type forwarding facades to point to the mono implementation (they have some way to ignore the version mismatch, not sure of the specifics).  

.NET Framework has its own implementation in System.ServiceModel.dll, we just provide the type forwarding facades for it if they are needed. Everything else provides its own facades (e.g. mono) and implementation. Every other netstandard2.0 support platform does the same (e.g. Unity).   

Previously in the 4.10.3 version, we only depended on netstandard2.0 api's. While each of the packages you would directly depend on (eg Primitives) had some complicated package dependencies which is a whole other topic as to why, the implementation all existed in the System.Private.ServiceModel package. If you look inside that, there's only a netstandard2.0 binary. The only platform that this code was actually executed on was .NET [Core], as every other platform has its own implementation. It was getting frustrating being limited to only netstandard2.0 apis. We couldn't make use of Span based api's for improved performance. Some api's have been improved, but we couldn't take advantage of them (e.g. a lot of string handling api's are a lot better now). The only reason we were limited to this is to have our assembly only depend on netstandard2.0 api's so that it could sit in the lib/netstandard2.0 folder.  

Then along came .NET Maui. While .NET Maui uses netX.0 based tfm's (e.g. net8.0-android), the set of tfm's it would search for included some xamarin tfm's. These take precedence over netstandard2.0. This resulted in a .NET Maui app wanting to use WCF finding one of the xamarin based tfm folders which suppresses picking up the runtime assembly. This meant for .NET Maui, we needed to add explicit net6.0 support. The packages were getting bloated with duplicate items.  

On top of that, there used to be a policy that we would harvest binaries from older releases and pull implementations from them into the latest package. So if you look at the 4.10.3 package, you'll find a lib/netcore50 from before netstandard was created, and corresponds to .NET Core 1.0. There's ref folders for netstandard 1.0, 1.1, and 1.3. This process was called package harvesting, and was intended to enable people pulling in the latest package and having things still work even if using a really old many years out of support version of .NET Core. You'll find this same thing in older .NET runtime packages too. At some point the dotnet team decided that enabling using newer packages on old out of support runtimes wasn't something that engineering effort should be spent on maintaining so this stopped. We didn't get the memo and I discovered they had made this change some time after it happened, so WCF lagged a bit on removing support for older out of support runtimes.  

Because of this policy change of not putting in effort to enable out of support scenarios, it made things simpler for us. I was advised that with the current state of the eco system it was fine for us to support only currently in support .NET versions and we didn't even need to worry about .NET Framework as everyone was multi-targeting now. Dropping .NET Framework support turned out to be a mistake as there were scenarios we weren't aware of. For example, powershell ships our assemblies in a standalone distribution (you don't need the .NET shared runtime installed to use it). One of their very important scenarios is using a powershell module that contains compiled assemblies that was compiled against .NET Framework and intended to run on the powershell which ships with Windows (and so uses .NET Framework). Those assemblies reference System.ServiceModel.dll to find WCF types. So we've added back a System.ServiceModel.dll shim for that scenario so that they can consume our latest packages and still run modules compiled for .NET Framework.  

Basically what this comes down to is you can still target netstandard2.0 and consume our latest packages and they will work on all platforms which support netstandard2.0, except out of support versions of .NET, ie .NET 6. If there's enough noise about libraries/packages needing to reference our assemblies that have a requirement to be able to run on out of support versions of the .NET runtime, we could consider adding back some package harvesting type of situation going back a limited amount of time, e.g. the previous LTS (so WCF Client 10.0 for example could harvest the last 8.x version). Currently we don't intend on doing this, but it's a potential option if this affects a lot of people.

**Comment by mconnew** [comment]
I forgot to explicitly address why we support netstandard2.0 instead of just being multi-targeted. There are products which have a plugin like extensibility which have an older implementation that runs on .NET Framework, and a newer implementation that runs on .NET. They have a single nestandard2.0 extensibility api which allows plugin authors to provide a single netstandard targeted implementation. They will run the same plugin on both the .NET Framework implementation, and the .NET implementation. The current model of supporting different platforms such as .NET Maui on iOS and Android is to use the same .NET runtime on all the platforms, but switch out the CLR pieces (garbage collector/heap management, JIT vs interpreter etc). So netstandard2.0 is becoming more a common api between .NET Framework and .NET than the broad landscape of many runtime implementations. It's why they didn't evolve netstandard beyond 2.1, because it was becoming too much work, and required too much coordination between various runtime maintainers to evolve the api surface forward. Even .NET Framework stopped at netstandard2.0. There isn't going to be any new runtimes created that implement netstandard2.0, anything new will take .NET and replace components of the CLR as needed and will support the .NET tfm's. So any existing netstandard2.0 platforms will have their own implementation (we aren't going to evolve the netstandard2.0 api surface), or it's one of .NET Framework or .NET.

**Comment by aueelis** [comment]
@mconnew This does indeed seem to be a much more complex situation than I expected. Thank you very much for the detailed write up!

**Comment by JonSaffron** [comment]
Yes, thanks for the detailed explanation. It does unfortunately make complete sense, whilst still leaving quite an annoying problem. I'd be tempted to open an issue against nuget instead, but I'm not sure where to even begin.
